src/header/common/origin.rs

```diff
@@ -0,0 +1,127 @@
+use header::{Header, Host, HeaderFormat};
+use std::fmt;
+use std::str::FromStr;
+use header::parsing::from_one_raw_str;
+
+/// The `Origin` header.
+///
+/// The `Origin` header is a version of the `Referer` header that is used for all HTTP fetches and `POST`s whose CORS flag is set.
+/// This header is often used to inform recipients of the security context of where the request was initiated.
+///
+///
+/// Following the spec, https://fetch.spec.whatwg.org/#origin-header, the value of this header is composed of
+/// a String (scheme), header::Host (host/port)
+///
+/// # Examples
+/// ```
+/// use hyper::header::{Headers, Origin, Host};
+///
+/// let mut headers = Headers::new();
+/// headers.set(
+///     Origin{
+///           scheme: "http".to_owned(),
+///           host: Host{hostname: "hyper.rs".to_owned(), port: None}
+///    }
+/// );
+/// ```
+/// ```
+/// use hyper::header::{Headers, Origin, Host};
+///
+/// let mut headers = Headers::new();
+/// headers.set(
+///     Origin{
+///           scheme: "https".to_owned(),
+///           host: Host{hostname: "wikipedia.org".to_owned(), port: Some(443)}
+///     }
+/// );
+/// ```
+
+#[derive(Clone, Debug)]
+pub struct Origin {
+    /// The scheme, such as http or https
+    pub scheme: String,
+    /// The host, such as Host{hostname: "hyper.rs".to_owned(), port: None}
+    pub host: Host,
+}
+
+impl Header for Origin {
+    fn header_name() -> &'static str {
+        static NAME: &'static str = "Origin";
+        NAME
+    }
+
+    fn parse_header(raw: &[Vec<u8>]) -> ::Result<Origin> {
+        from_one_raw_str(raw).and_then(|s: String| {
+            Origin::from_str(&s)
+        })
+    }
+}
+
+impl FromStr for Origin {
+    type Err = ::Error;
+
+    fn from_str(s: &str) -> ::Result<Origin> { // not exactly sure why this works and why im not getting "you need 2 types error"
+        let idx = match s.find("://") {
+            Some(idx) => idx,
+            None => return Err(::Error::Header)
+        };
+        // idx + 3 because thats how long "://" is
+        let (scheme, etc) = (&s[..idx], &s[idx + 3..]);
+        let host = try!(Host::from_str(etc));
+
+
+        Ok(Origin{
+            scheme: scheme.to_owned(),
+            host: host
+        })
+    }
+}
+
+impl HeaderFormat for Origin {
+    fn fmt_header(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        // I wanna the responsibilty of writing the host/port to Host
```

Comment:
That doesn't sound right. I meant for this `write!` to be the only thing in the function. That way the `Result` is returned.